------------------------
Author: "Eddie Grinman"
Date: "January 23rd, 2019"
Notes: Reading the data, Principle Component Analysis, Differential Expression Analysis,
  Hierarchical Clustering Heatmaps, Volcano plots, Gene Biotype Analysis, Pathway
  Analysis
Output: html_document
Title: RNASeq Workflow in R
------------------------



The following snippet installs packages that are needed. You can modify this part to include any packages you want to install.
If the packages are already installed, nothing will happen. 
These may not be exhaustive of all the packages that are required for this pipeline, but can be used as a starting point for automatic installation. 
```{r warning=FALSE}
needed_packages<-c("ggplot2","ggrepel", "ggdendro","dplyr", "plyr", "stats","grid","reshape","DESeq2","biomaRt","ReactomePA","devtools","rgl","FactoMineR","factoextra","org.Mm.eg.db","biomaRt","GOstats","GO.db", "pheatmap", "GenomicFeatures")
already_installed_packages<-rownames(installed.packages())
need_to_install<-setdiff(needed_packages,already_installed_packages)
if (length(need_to_install)>0){
  source("https://bioconductor.org/biocLite.R")
  biocLite(need_to_install)
}
```

```{r warning=FALSE, message=FALSE}
library(devtools)
library(rgl)
library(FactoMineR)
library(stats)
library(factoextra)
library(grid)
library(ggrepel)
require(dplyr)
library(plyr)
library(pheatmap)
```


I have a set of raw count data from an RNAseq dataset (Raveendra, et al., 2018, PNAS). 
This dataset has 2 conditions (KD and Control) with N=3 per group. 
I start by changing to the directory that contains the data, and importing the csv file of the raw counts with sample name on the column side and gene name on the row side. 

## Import Raw Counts
```{r}
setwd("/Users/eddiegrinman/Desktop/Lab_data/RNAseq/Raveendra_2018")
input <- read.csv("raw_counts.csv")
row.names(input)<- input$Name #I'm adding the gene names as row names
input <- input[,-1] #I'm removing the gene names as a variable from the data frame, because otherwise it will mess up DEseq. 
head(input)
```

# Principle Component Analysis
```{r warning=FALSE}
library(devtools)
library(rgl)
library(FactoMineR)
library(stats)
library(factoextra)

data_all <- read.csv('TPM_counts.csv')
rownames(data_all) <- data_all$Name
data_all$Name <- NULL
data_all_numeric <- as.data.frame(t(data_all)) #Transpose dataframe
new_data <- data_all_numeric #Now it's called new_data
names(new_data) <- rownames(new_data) #Give the new_data the stored rowname
new_active <- new_data
```

## Write a function to define group
```{r}
getgroup <- function(x){
  if(grepl('KD',x)){
    return("KD")
  }
  if(grepl('Control',x)){
    return("Control")
  }
}
data_all_patients <- colnames(data_all) #Get the list of patient identification
group_all_patients <- sapply(data_all_patients, getgroup) #Apply getgroup function to the list of patient identification
group_all_patients_vector <- unlist(group_all_patients, use.names = FALSE) #Transform list to vector
```

## 2d PCA plot
```{r}
new_active_pca <- prcomp(new_active) 
#scale = TRUE) #Use prcomp() for PCA
summary(new_active_pca)
```

## Coordinates of Samples on the principal components
```{r}
ind_coord_new_active_pca <- new_active_pca$x
head(ind_coord_new_active_pca[,1:4])
```

## Sample factor map
```{r}
fviz_pca_ind(new_active_pca) + theme_minimal()
```

## Color Samples by goups
```{r message=FALSE}
fviz_pca_ind(new_active_pca, 
             habillage = group_all_patients_vector, addEllipses = TRUE, ellipse.level=0.95) +
  theme_minimal()
```

## Differential Expression Analysis
```{r message=FALSE, warning=FALSE}
library(DESeq2)
colD<-data.frame(condition=as.factor(c(rep("KD",3),rep("Control",3))))
#This will lump each of the 3 groups together, by name. IE. KD 1-3 will be in the KD group.

colnames(input) <- as.factor(c(rep("KD",3),rep("Control",3)))
input2 <- as.matrix(input)
storage.mode(input2) = "integer"
dds <- DESeqDataSetFromMatrix(countData = input2, colData = colD, design = ~ condition)
dds <- DESeq(dds)
dds <- estimateSizeFactors(dds) 
normcounts<-counts(dds, normalized=TRUE) #Making Normalized Counts
head(normcounts)

KDvControl <- as.data.frame(results(dds, contrast = c("condition","KD","Control"))) #This will perform differential expression between KD and Control groups

KDvControl$nlog10padj <- -log10(KDvControl$padj) #Adding -log10 p-adj values for later volcano plotting

KDvControl$Gene <- rownames(KDvControl)  #Making the gene names as a variable in the data frame for easy merging with norm_counts dataframe

head(KDvControl)
```


# Gene Biotype Analysis
```{r}
library(biomaRt)
ensembl <- useEnsembl(biomart="ensembl")
mousegenes <- useEnsembl(biomart="ensembl", dataset='mmusculus_gene_ensembl') #Accessing mouse gene database from Ensembl
```


## Getting the gene information for the significant genes in each comparison
```{r}
gene_biotypes <- getBM(attributes=c('external_gene_name','gene_biotype','ensembl_gene_id', 'ensembl_transcript_id_version', 'entrezgene'),filters = 'ensembl_transcript_id_version',values = KDvControl$Gene, mart = mousegenes) #Salmon output gives the ensembl transcript IDs which we have to match back to the external gene name
colnames(gene_biotypes)[4]<-'Gene'
KDvControl <- merge(KDvControl,gene_biotypes,by="Gene")

KDvControl$sqnlog10p <- sqrt(KDvControl$nlog10padj) #This will be the y-axis of the volcano plot
head(KDvControl)
```

## Plot the volcano plot
```{r}
library(ggplot2)
g <- ggplot(KDvControl, aes(x=log2FoldChange,y=sqnlog10p))
g+geom_point(color='grey')+
  geom_hline(mapping=NULL, yintercept=sqrt(-log10(0.05)),size=0.5,linetype='dotted') +
  geom_vline(mapping=NULL, xintercept=0,size=0.5, linetype='dotted') +
  scale_color_manual(values = c("grey"))+
  theme_bw(base_size = 20) + theme(legend.position = "bottom")+
  theme_bw(base_size = 20) + theme(legend.position = "bottom")+
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))+
  scale_y_continuous(name='SqrRT-Log10 pAdj')+
  scale_x_continuous(name='Log2 Fold Change (KD vs Control)')+
  ggtitle(label = '', subtitle = 'Differential Gene Expression between KD and Control')
```


# Heatmap
```{r}
library(pheatmap)
KD_heatmap <- subset(KDvControl,baseMean>0) 
KD_heatmap <- subset(KD_heatmap,padj<0.05) #Significant genes only
x <- as.data.frame(subset(normcounts,rownames(normcounts) %in% KD_heatmap$Gene)) #Subsetting all the norm counts by the genes that are differentially expressed in one group
cal_z_score <- function(x){
  (x - mean(x)) / sd(x)
}
data_subset_norm <- t(apply(as.matrix(x), 1, cal_z_score)) #Applying a row Z-score transformation to the normalized count data
head(data_subset_norm)
pheatmap(data_subset_norm, main = 'Differentially Expressed Genes', show_rownames = FALSE, show_colnames = TRUE) #Plot heat map
```

## Next, I would like to plot differentially expressed noncoding RNAs in this data set as a heatmap. 
```{r}
ncgenes <- subset(KD_heatmap, !(KD_heatmap$gene_biotype == 'protein_coding'))
x <- as.data.frame(subset(normcounts,rownames(normcounts) %in% ncgenes$Gene)) #Subsetting all the norm counts by the noncoding RNA genes
data_subset_norm <- t(apply(as.matrix(x), 1, cal_z_score)) #Applying a row Z-score transformation to the normalized count data
pheatmap(data_subset_norm, main = 'Differentially Expressed Noncoding RNAs', show_rownames = TRUE, show_colnames = TRUE) #Plot heat map
```



## I would also like to label these ncRNAs on the volcano plot made previously. 
```{r warning=FALSE}
KDvControl$significant <- ifelse(grepl(paste(ncgenes$Gene,collapse='|'), KDvControl$Gene),"Blue", "Grey") #Making a new column that marks all the significant non-coding genes as 'Blue' and the remaining genes as 'Grey'
KDvControl <- KDvControl[rev(order(as.numeric(factor(KDvControl$significant)))),] #Reordering the data frame so that the labeled genes will be plotted in front of the grey
KDvControl$sqnlog10p <- sqrt(KDvControl$nlog10padj) #This is the y axis of the volcano plot

g <- ggplot(KDvControl, aes(x=log2FoldChange,y=sqnlog10p))
g+geom_point(color='grey')+
  geom_point(color=KDvControl$significant)+
  geom_hline(mapping=NULL, yintercept=sqrt(-log10(0.05)),size=0.5,linetype='dotted') +
  geom_vline(mapping=NULL, xintercept=0,size=0.5, linetype='dotted') +
  scale_color_manual(values = c("grey"))+
  theme_bw(base_size = 20) + theme(legend.position = "bottom")+
  theme_bw(base_size = 20) + theme(legend.position = "bottom")+
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))+
  scale_y_continuous(name='SqrRT-Log10 pAdj')+
  scale_x_continuous(name='Log2 Fold Change (KD vs Control)')+
  ggtitle(label = '', subtitle = 'Differentially Expressed non-coding RNAs')
```



# Reactome Pathway Analysis
```{r}
require(ReactomePA)
library(org.Mm.eg.db)
library(reactome.db)

downregulated <- subset(KD_heatmap, log2FoldChange < 0) #As an example, I am performing pathway analysis on only the downregulated genes
geneList <- downregulated$entrezgene #The entrez gene is the input for Reactome pathway analysis
```

```{r}
x <- enrichPathway(gene=geneList, organism="mouse", pvalueCutoff=0.05, readable=T, maxGSSize = 2000, minGSSize = 2)
barplot(x,showCategory = 20) #barplot showing pathways with the number of genes in each pathway on the x-axis and heat to show signifiance
emapplot(x, showCategory = 10) #network plot to show how each pathway relates to others in terms of overlapping genes

cnetplot(x, categorySize="pvalue", foldChange=geneList) #Network plot showing each gene that is contributing to a given pathway
```




